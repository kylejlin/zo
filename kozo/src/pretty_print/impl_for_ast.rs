use super::*;

use crate::syntax_tree::ast::*;

impl Display for PrettyPrinted<'_, Expr> {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        fmt_expr(self.0, f, Indentation { soft_tab_count: 0 })
    }
}

// Every `fmt_{node}` function writes an indent
// at the beginning.
// It does _not_ write a newline at the end.

fn fmt_expr(expr: &Expr, f: &mut Formatter<'_>, indent: Indentation) -> FmtResult {
    match expr {
        Expr::Ind(e) => fmt_ind(e, f, indent),
        Expr::Vcon(e) => fmt_vcon(e, f, indent),
        Expr::Match(e) => fmt_match(e, f, indent),
        Expr::Fun(e) => fmt_fun(e, f, indent),
        Expr::App(e) => fmt_app(e, f, indent),
        Expr::For(e) => fmt_for(e, f, indent),
        Expr::Deb(e) => fmt_deb(e, f, indent),
        Expr::Universe(e) => fmt_universe(e, f, indent),
    }
}

fn fmt_ind(ind: &RcSemHashed<Ind>, f: &mut Formatter<'_>, indent: Indentation) -> FmtResult {
    let i1 = indent.incremented();
    let universe_level = ind.value.universe_level.0;
    write!(f, "{indent}(\n{i1}ind\n{i1}Type{universe_level}\n")?;
    fmt_str_literal(ind.value.name.clone(), f, i1)?;
    write!(f, "\n")?;

    fmt_parenthesized_expressions(ind.value.index_types.clone(), f, i1)?;
    write!(f, "\n")?;

    fmt_parenthesized_vcon_defs(ind.value.vcon_defs.clone(), f, i1)?;
    write!(f, "\n{indent})")?;

    Ok(())
}

fn fmt_str_literal(
    str_literal: Rc<StringValue>,
    f: &mut Formatter<'_>,
    indent: Indentation,
) -> FmtResult {
    todo!()
}

fn fmt_parenthesized_vcon_defs(
    parenthesized_vcon_defs: RcSemHashed<Box<[VconDef]>>,
    f: &mut Formatter<'_>,
    indent: Indentation,
) -> FmtResult {
    todo!()
}

fn fmt_vcon(vcon: &RcSemHashed<Vcon>, f: &mut Formatter<'_>, indent: Indentation) -> FmtResult {
    todo!()
}

fn fmt_match(m: &RcSemHashed<Match>, f: &mut Formatter<'_>, indent: Indentation) -> FmtResult {
    todo!()
}

fn fmt_fun(fun: &RcSemHashed<Fun>, f: &mut Formatter<'_>, indent: Indentation) -> FmtResult {
    todo!()
}

fn fmt_app(app: &RcSemHashed<App>, f: &mut Formatter<'_>, indent: Indentation) -> FmtResult {
    todo!()
}

fn fmt_for(for_: &RcSemHashed<For>, f: &mut Formatter<'_>, indent: Indentation) -> FmtResult {
    todo!()
}

fn fmt_deb(deb: &RcSemHashed<DebNode>, f: &mut Formatter<'_>, indent: Indentation) -> FmtResult {
    let index = deb.value.deb.0;
    write!(f, "{indent}Type{index}")
}

fn fmt_universe(
    universe: &RcSemHashed<UniverseNode>,
    f: &mut Formatter<'_>,
    indent: Indentation,
) -> FmtResult {
    let level = universe.value.level.0;
    write!(f, "{indent}Type{level}")
}

fn fmt_parenthesized_expressions(
    parenthesized_expressions: RcSemHashed<Box<[Expr]>>,
    f: &mut Formatter<'_>,
    indent: Indentation,
) -> FmtResult {
    todo!()
}
