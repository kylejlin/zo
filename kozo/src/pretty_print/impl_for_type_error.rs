use super::*;

use crate::{
    syntax_tree::{ipist::rc_hashed, ipist_to_ast::*},
    typecheck::TypeError,
};

impl Display for PrettyPrint<'_, TypeError> {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        match &self.0 {
            TypeError::InvalidDeb { deb, tcon_len } => f
                .debug_struct("TypeError::InvalidDeb")
                .field("deb", &deb.value)
                .field("tcon_len", tcon_len)
                .finish(),

            TypeError::InvalidVconIndex(vcon) => {
                let mut converter = IpistToAstConverter::default();
                let vcon_ast = converter.convert_vcon(rc_hashed(vcon.clone()));
                f.debug_struct("TypeError::InvalidVconIndex")
                    .field("vcon", &vcon_ast.pretty_printed())
                    .finish()
            }

            TypeError::UnexpectedNonTypeExpression { expr, type_ } => {
                let mut converter = IpistToAstConverter::default();
                let expr_ast = converter.convert(expr.clone());
                f.debug_struct("TypeError::UnexpectedNonTypeExpression")
                    .field("expr", &expr_ast.pretty_printed())
                    .field("type_", &type_.raw().pretty_printed())
                    .finish()
            }

            TypeError::UniverseInconsistencyInIndDef {
                index_or_param_type,
                level,
                ind,
            } => {
                let mut converter = IpistToAstConverter::default();
                let index_or_param_type_ast = converter.convert(index_or_param_type.clone());
                let ind_ast = converter.convert_ind(rc_hashed(ind.clone()));
                f.debug_struct("TypeError::UniverseInconsistencyInIndDef")
                    .field(
                        "index_or_param_type",
                        &index_or_param_type_ast.pretty_printed(),
                    )
                    .field("level", &level)
                    .field("ind", &ind_ast.pretty_printed())
                    .finish()
            }

            TypeError::WrongNumberOfIndexArguments {
                def,
                expected,
                actual,
            } => {
                let mut converter = IpistToAstConverter::default();
                let def_ast = converter.convert_vcon_def(def.clone());
                f.debug_struct("TypeError::WrongNumberOfIndexArguments")
                    .field("def", &def_ast.pretty_printed())
                    .field("expected", expected)
                    .field("actual", actual)
                    .finish()
            }

            TypeError::NonInductiveMatcheeType { expr, type_ } => {
                let mut converter = IpistToAstConverter::default();
                let expr_ast = converter.convert(expr.clone());
                f.debug_struct("TypeError::NonInductiveMatcheeType")
                    .field("expr", &expr_ast.pretty_printed())
                    .field("type_", &type_.raw().pretty_printed())
                    .finish()
            }

            TypeError::WrongNumberOfMatchCases {
                match_,
                matchee_type_ind,
            } => {
                let mut converter = IpistToAstConverter::default();
                let match_ast = converter.convert_match(rc_hashed(match_.clone()));
                f.debug_struct("TypeError::WrongNumberOfMatchCases")
                    .field("match_", &match_ast.pretty_printed())
                    .field(
                        "matchee_type_ind",
                        &rc_hashed(matchee_type_ind.raw().clone()).pretty_printed(),
                    )
                    .finish()
            }

            TypeError::WrongMatchCaseArity {
                actual_node,
                expected,
                match_,
                match_case_index,
            } => {
                let mut converter = IpistToAstConverter::default();
                let match_ast = converter.convert_match(rc_hashed(match_.clone()));
                f.debug_struct("TypeError::WrongMatchCaseArity")
                    .field("actual_node", &actual_node.value)
                    .field("expected", expected)
                    .field("match_", &match_ast.pretty_printed())
                    .field("match_case_index", match_case_index)
                    .finish()
            }

            TypeError::IllegallyDismissedMatchCase {
                match_,
                match_case_index,
            } => {
                let mut converter = IpistToAstConverter::default();
                let match_ast = converter.convert_match(rc_hashed(match_.clone()));
                f.debug_struct("TypeError::IllegallyDismissedMatchCase")
                    .field("match_", &match_ast.pretty_printed())
                    .field("match_case_index", match_case_index)
                    .finish()
            }

            TypeError::TypeMismatch {
                expr,
                expected_type,
                actual_type,
            } => {
                let mut converter = IpistToAstConverter::default();
                let expr_ast = converter.convert(expr.clone());
                f.debug_struct("TypeError::TypeMismatch")
                    .field("expr", &expr_ast.pretty_printed())
                    .field("expected_type", &expected_type.raw().pretty_printed())
                    .field("actual_type", &actual_type.raw().pretty_printed())
                    .finish()
            }

            TypeError::CalleeTypeIsNotAForExpression { app, callee_type } => {
                let mut converter = IpistToAstConverter::default();
                let app_ast = converter.convert_app(rc_hashed(app.clone()));
                f.debug_struct("TypeError::CalleeTypeIsNotAForExpression")
                    .field("app", &app_ast.pretty_printed())
                    .field("callee_type", &callee_type.raw().pretty_printed())
                    .finish()
            }

            TypeError::WrongNumberOfAppArguments {
                app,
                callee_type,
                expected,
                actual,
            } => {
                let mut converter = IpistToAstConverter::default();
                let app_ast = converter.convert_app(rc_hashed(app.clone()));
                f.debug_struct("TypeError::WrongNumberOfAppArguments")
                    .field("app", &app_ast.pretty_printed())
                    .field(
                        "callee_type",
                        &rc_hashed(callee_type.raw().clone()).pretty_printed(),
                    )
                    .field("expected", expected)
                    .field("actual", actual)
                    .finish()
            }
        }
    }
}
