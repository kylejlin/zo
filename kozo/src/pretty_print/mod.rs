use std::fmt::{Display, Formatter, Result as FmtResult};

pub const SOFT_TAB_SIZE: usize = 4;

#[derive(Clone, Debug, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct PrettyPrinted<'a, T>(pub &'a T);

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
struct SoftTabCount(pub usize);

fn get_indent_str(count: SoftTabCount) -> String {
    " ".repeat(count.0 * SOFT_TAB_SIZE)
}

mod impl_for_ast;
