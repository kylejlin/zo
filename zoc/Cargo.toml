[package]
name = "zoc"
version = "0.1.0"
edition = "2021"

[dependencies]
rustc-hash = "1.1.0"

[dev-dependencies]
pretty_assertions = "1.3.0"
insta = "1.29.0"

[build-dependencies]
kiki = "7"
walkdir = "2"
sha256 = "1.1.4"
