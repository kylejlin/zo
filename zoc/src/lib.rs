pub mod check_erasability;
pub mod eval;
pub mod hash;
pub mod pretty_print;
pub mod syntax_tree;
pub mod typecheck;

#[cfg(test)]
pub mod test_utils;
