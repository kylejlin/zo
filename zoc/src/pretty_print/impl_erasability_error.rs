use super::*;

use crate::check_erasability::ErasabilityError;

impl Display for PrettyPrint<'_, ErasabilityError> {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        match &self.0 {
            ErasabilityError::MatcheeTypeTypeIsErasableButReturnTypeTypeIsNotErasable {
                match_,
                matchee_type_type,
                match_return_type_type,
            } => f
                .debug_struct(
                    "ErasabilityError::MatcheeTypeTypeIsErasableButReturnTypeTypeIsNotErasable",
                )
                .field("match_", &match_.pretty_printed())
                .field("matchee_type_type", &matchee_type_type.pretty_printed())
                .field(
                    "match_return_type_type",
                    &match_return_type_type.pretty_printed(),
                )
                .finish(),
        }
    }
}
