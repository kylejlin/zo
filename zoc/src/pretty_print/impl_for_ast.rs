use super::*;

use crate::syntax_tree::ast::*;

impl Display for PrettyPrint<'_, Expr> {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        fmt_expr(self.0.clone(), f, Indentation { soft_tab_count: 0 })
    }
}

impl Display for PrettyPrint<'_, RcHashed<Ind>> {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        fmt_ind(self.0.clone(), f, Indentation { soft_tab_count: 0 })
    }
}

impl Display for PrettyPrint<'_, RcHashedVec<VconDef>> {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        fmt_parenthesized_vcon_defs(self.0.clone(), f, Indentation { soft_tab_count: 0 })
    }
}

impl Display for PrettyPrint<'_, VconDef> {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        fmt_vcon_def(self.0, f, Indentation { soft_tab_count: 0 })
    }
}

impl Display for PrettyPrint<'_, RcHashed<Vcon>> {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        fmt_vcon(self.0.clone(), f, Indentation { soft_tab_count: 0 })
    }
}

impl Display for PrettyPrint<'_, RcHashed<Match>> {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        fmt_match(self.0.clone(), f, Indentation { soft_tab_count: 0 })
    }
}

impl Display for PrettyPrint<'_, RcHashedVec<MatchCase>> {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        fmt_parenthesized_match_cases(self.0.clone(), f, Indentation { soft_tab_count: 0 })
    }
}

impl Display for PrettyPrint<'_, MatchCase> {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        fmt_match_case(self.0, f, Indentation { soft_tab_count: 0 })
    }
}

impl Display for PrettyPrint<'_, RcHashed<Fun>> {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        fmt_fun(self.0.clone(), f, Indentation { soft_tab_count: 0 })
    }
}

impl Display for PrettyPrint<'_, RcHashed<App>> {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        fmt_app(self.0.clone(), f, Indentation { soft_tab_count: 0 })
    }
}

impl Display for PrettyPrint<'_, RcHashed<For>> {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        fmt_for(self.0.clone(), f, Indentation { soft_tab_count: 0 })
    }
}

impl Display for PrettyPrint<'_, RcHashed<DebNode>> {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        fmt_deb(self.0.clone(), f, Indentation { soft_tab_count: 0 })
    }
}

impl Display for PrettyPrint<'_, RcHashed<UniverseNode>> {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        fmt_universe(self.0.clone(), f, Indentation { soft_tab_count: 0 })
    }
}

// Every `fmt_{node}` function writes an indent
// at the beginning.
// It does _not_ write a newline at the end.

fn fmt_expr(expr: Expr, f: &mut Formatter<'_>, indent: Indentation) -> FmtResult {
    match expr {
        Expr::Ind(e) => fmt_ind(e, f, indent),
        Expr::Vcon(e) => fmt_vcon(e, f, indent),
        Expr::Match(e) => fmt_match(e, f, indent),
        Expr::Fun(e) => fmt_fun(e, f, indent),
        Expr::App(e) => fmt_app(e, f, indent),
        Expr::For(e) => fmt_for(e, f, indent),
        Expr::Deb(e) => fmt_deb(e, f, indent),
        Expr::Universe(e) => fmt_universe(e, f, indent),
    }
}

fn fmt_ind(ind: RcHashed<Ind>, f: &mut Formatter<'_>, indent: Indentation) -> FmtResult {
    let i1 = indent.incremented();
    let universe_level = ind.hashee.universe_level.0;
    write!(f, "{indent}(\n{i1}ind\n{i1}Type{universe_level}\n")?;
    fmt_str_literal(ind.hashee.name.clone(), f, i1)?;
    write!(f, "\n")?;

    fmt_parenthesized_expressions(ind.hashee.index_types.clone(), f, i1)?;
    write!(f, "\n")?;

    fmt_parenthesized_vcon_defs(ind.hashee.vcon_defs.clone(), f, i1)?;
    write!(f, "\n{indent})")?;

    Ok(())
}

fn fmt_str_literal(
    str_literal: Rc<StringValue>,
    f: &mut Formatter<'_>,
    indent: Indentation,
) -> FmtResult {
    write!(f, "{indent}\"")?;

    for c in str_literal.0.chars() {
        if c.is_ascii_alphanumeric() || " _`~!@#$%^&*()-_=+[]|;:',<.>/?".contains(c) {
            write!(f, "{c}")?;
        } else {
            write!(f, "{{0x{:X}}}", u32::from(c))?;
        }
    }

    write!(f, "\"")?;
    Ok(())
}

fn fmt_parenthesized_vcon_defs(
    defs: RcHashedVec<VconDef>,
    f: &mut Formatter<'_>,
    indent: Indentation,
) -> FmtResult {
    if defs.hashee.is_empty() {
        return write!(f, "{indent}()");
    }

    write!(f, "{indent}(")?;
    let i1 = indent.incremented();

    for def in defs.hashee.iter() {
        write!(f, "\n")?;
        fmt_vcon_def(def, f, i1)?;
    }

    write!(f, "\n{indent})")?;
    Ok(())
}

fn fmt_vcon_def(def: &VconDef, f: &mut Formatter<'_>, indent: Indentation) -> FmtResult {
    write!(f, "{indent}(\n")?;

    let i1 = indent.incremented();
    fmt_parenthesized_expressions(def.param_types.clone(), f, i1)?;
    write!(f, "\n")?;

    fmt_parenthesized_expressions(def.index_args.clone(), f, i1)?;

    write!(f, "\n{indent})")?;
    Ok(())
}

fn fmt_vcon(vcon: RcHashed<Vcon>, f: &mut Formatter<'_>, indent: Indentation) -> FmtResult {
    let i1 = indent.incremented();
    write!(f, "{indent}(\n{i1}vcon\n")?;

    fmt_ind(vcon.hashee.ind.clone(), f, i1)?;

    let vcon_index = vcon.hashee.vcon_index;
    write!(f, "\n{i1}{vcon_index}\n{indent})")?;
    Ok(())
}

fn fmt_match(m: RcHashed<Match>, f: &mut Formatter<'_>, indent: Indentation) -> FmtResult {
    let i1 = indent.incremented();
    write!(f, "{indent}(\n{i1}match\n")?;

    fmt_expr(m.hashee.matchee.clone(), f, i1)?;
    write!(f, "\n")?;

    let return_type_arity = m.hashee.return_type_arity;
    write!(f, "{i1}{return_type_arity}\n")?;

    fmt_expr(m.hashee.return_type.clone(), f, i1)?;
    write!(f, "\n")?;

    fmt_parenthesized_match_cases(m.hashee.cases.clone(), f, i1)?;
    write!(f, "\n{indent})")?;
    Ok(())
}

fn fmt_parenthesized_match_cases(
    cases: RcHashedVec<MatchCase>,
    f: &mut Formatter<'_>,
    indent: Indentation,
) -> FmtResult {
    if cases.hashee.is_empty() {
        return write!(f, "{indent}()");
    }

    write!(f, "{indent}(")?;
    let i1 = indent.incremented();

    for case in cases.hashee.iter() {
        write!(f, "\n")?;
        fmt_match_case(case, f, i1)?;
    }

    write!(f, "\n{indent})")?;
    Ok(())
}

fn fmt_match_case(case: &MatchCase, f: &mut Formatter<'_>, indent: Indentation) -> FmtResult {
    let i1 = indent.incremented();
    let case_arity = case.arity;
    write!(f, "{indent}(\n{i1}{case_arity}\n")?;

    fmt_expr(case.return_val.clone(), f, i1)?;
    write!(f, "\n{indent})")?;
    Ok(())
}

fn fmt_fun(fun: RcHashed<Fun>, f: &mut Formatter<'_>, indent: Indentation) -> FmtResult {
    let i1 = indent.incremented();
    write!(f, "{indent}(\n{i1}fun\n")?;

    fmt_decreasing_index(fun.hashee.decreasing_index, f, i1)?;
    write!(f, "\n")?;

    fmt_parenthesized_expressions(fun.hashee.param_types.clone(), f, i1)?;
    write!(f, "\n")?;

    fmt_expr(fun.hashee.return_type.clone(), f, i1)?;
    write!(f, "\n")?;

    fmt_expr(fun.hashee.return_val.clone(), f, i1)?;
    write!(f, "\n{indent})")?;
    Ok(())
}

fn fmt_decreasing_index(
    index: Option<usize>,
    f: &mut Formatter<'_>,
    indent: Indentation,
) -> FmtResult {
    match index {
        Some(index) => write!(f, "{indent}{index}"),
        None => write!(f, "{indent}nonrec"),
    }
}

fn fmt_app(app: RcHashed<App>, f: &mut Formatter<'_>, indent: Indentation) -> FmtResult {
    if app.hashee.args.hashee.is_empty() {
        write!(f, "{indent}(")?;
        fmt_expr(app.hashee.callee.clone(), f, indent)?;
        write!(f, ")")?;
        return Ok(());
    }

    write!(f, "{indent}(\n")?;

    let i1 = indent.incremented();
    fmt_expr(app.hashee.callee.clone(), f, i1)?;

    for arg in app.hashee.args.hashee.iter() {
        write!(f, "\n")?;
        fmt_expr(arg.clone(), f, i1)?;
    }

    write!(f, "\n{indent})")?;
    Ok(())
}

fn fmt_for(for_: RcHashed<For>, f: &mut Formatter<'_>, indent: Indentation) -> FmtResult {
    let i1 = indent.incremented();
    write!(f, "{indent}(\n{i1}for\n")?;

    fmt_parenthesized_expressions(for_.hashee.param_types.clone(), f, i1)?;
    write!(f, "\n")?;

    fmt_expr(for_.hashee.return_type.clone(), f, i1)?;
    write!(f, "\n{indent})")?;
    Ok(())
}

fn fmt_deb(deb: RcHashed<DebNode>, f: &mut Formatter<'_>, indent: Indentation) -> FmtResult {
    let index = deb.hashee.deb.0;
    write!(f, "{indent}{index}")
}

fn fmt_universe(
    universe: RcHashed<UniverseNode>,
    f: &mut Formatter<'_>,
    indent: Indentation,
) -> FmtResult {
    let level = universe.hashee.level.0;
    write!(f, "{indent}Type{level}")
}

fn fmt_parenthesized_expressions(
    parenthesized_expressions: RcHashedVec<Expr>,
    f: &mut Formatter<'_>,
    indent: Indentation,
) -> FmtResult {
    if parenthesized_expressions.hashee.is_empty() {
        return write!(f, "{indent}()");
    }

    write!(f, "{indent}(")?;
    let i1 = indent.incremented();

    for expr in parenthesized_expressions.hashee.iter() {
        write!(f, "\n")?;
        fmt_expr(expr.clone(), f, i1)?;
    }

    write!(f, "\n{indent})")?;
    Ok(())
}
