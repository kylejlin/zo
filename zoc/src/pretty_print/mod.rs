use std::{
    fmt::{Debug, Display, Formatter, Result as FmtResult},
    rc::Rc,
};

#[derive(Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct PrettyPrint<'a, T>(pub &'a T);

impl<'a, T> Debug for PrettyPrint<'a, T>
where
    PrettyPrint<'a, T>: Display,
{
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        Display::fmt(self, f)
    }
}

mod impl_for_ast;
mod impl_for_type_error;

pub trait PrettyPrinted {
    fn pretty_printed(&self) -> PrettyPrint<'_, Self>
    where
        Self: Sized;
}

impl<T> PrettyPrinted for T {
    fn pretty_printed(&self) -> PrettyPrint<'_, Self>
    where
        Self: Sized,
    {
        PrettyPrint(self)
    }
}

pub trait PrettyUnwrap {
    type Output;

    fn pretty_unwrap(self) -> Self::Output;
}

pub trait PrettyUnwrapErr {
    type Output;

    fn pretty_unwrap_err(self) -> Self::Output;
}

impl<T, E> PrettyUnwrap for Result<T, E>
where
    for<'a> PrettyPrint<'a, E>: Display,
{
    type Output = T;

    fn pretty_unwrap(self) -> T {
        match self {
            Ok(v) => v,
            Err(e) => {
                // We use the "#" flag so that if the
                // Display impl of `PrettyPrint<E>` uses
                // `Formatter:debug_struct` (or `debug_tuple`, etc.),
                // the output will be printed with easy-to-read
                // indentation.
                panic!(
                    "called `Result::unwrap()` on an `Err` value:\n{:#}",
                    PrettyPrint(&e)
                );
            }
        }
    }
}

impl<T, E> PrettyUnwrapErr for Result<T, E>
where
    for<'a> PrettyPrint<'a, T>: Display,
{
    type Output = E;

    fn pretty_unwrap_err(self) -> E {
        match self {
            Ok(v) => {
                panic!(
                    "called `Result::unwrap_err()` on an `Ok` value:\n{:#}",
                    PrettyPrint(&v)
                );
            }
            Err(e) => e,
        }
    }
}

pub const SOFT_TAB: &str = "    ";

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
struct Indentation {
    pub soft_tab_count: usize,
}

impl Indentation {
    pub fn incremented(self) -> Self {
        Self {
            soft_tab_count: self.soft_tab_count + 1,
        }
    }
}

impl Display for Indentation {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        for _ in 0..self.soft_tab_count {
            f.write_str(SOFT_TAB)?;
        }
        Ok(())
    }
}
