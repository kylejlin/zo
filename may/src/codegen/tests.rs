use super::*;

use zoc::pretty_print::PrettyPrint;

fn parse_or_panic(src: &str) -> mnode::Expr {
    let tokens = crate::lexer::lex(src).unwrap();
    crate::parser::parse(tokens).unwrap()
}

fn assert_expr_is_well_typed_under_empty_tcon(ast: znode::Expr) {
    use zoc::{
        eval::Normalized,
        pretty_print::PrettyUnwrap,
        syntax_tree::{lexer::lex as zo_lex, parser::parse as zo_parse},
        typecheck::{LazyTypeContext, TypeChecker},
    };

    let src = PrettyPrint(&ast).to_string();
    let tokens = zo_lex(&src).unwrap();
    let ost = zo_parse(tokens).unwrap();

    let empty = Normalized::<[_; 0]>::new();
    TypeChecker::default()
        .get_type(
            ost.into(),
            LazyTypeContext::Base(empty.as_ref().convert_ref()),
        )
        .pretty_unwrap();
}

#[test]
fn two() {
    let src = r#"
ind Nat
    case zero
    case succ(_: Nat)
    return Set0

succ(succ(zero))
"#;
    let cst = parse_or_panic(src);
    let zo = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(zo.clone());

    insta::assert_display_snapshot!(PrettyPrint(&zo));
}

#[test]
fn add_two_three() {
    let src = r#"
ind Nat
    case zero
    case succ(_: Nat)
    return Set0

let _2 = succ(succ(zero))

let _3 = succ(_2)

fun add(-a: Nat, b: Nat): Nat
    match a
    case zero:
        b
    case succ(a_pred):
        succ(add(a_pred, b))
    return1 Nat

add(_2, _3)
"#;
    let cst = parse_or_panic(src);
    let zo = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(zo.clone());

    insta::assert_display_snapshot!(PrettyPrint(&zo));
}

#[test]
fn list() {
    let src = r#"
ind(T: Set0) List
    case nil
    case cons(_: T, _: List)
    return Set0
List
"#;
    let cst = parse_or_panic(src);
    let zo = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(zo.clone());

    insta::assert_display_snapshot!(PrettyPrint(&zo));
}

#[test]
fn list_nil() {
    let src = r#"
ind(T: Set0) List
    case nil
    case cons(_: T, _: List)
    return Set0
nil
"#;
    let cst = parse_or_panic(src);
    let zo = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(zo.clone());

    insta::assert_display_snapshot!(PrettyPrint(&zo));
}

#[test]
fn list_cons() {
    let src = r#"
ind(T: Set0) List
    case nil
    case cons(_: T, _: List)
    return Set0
cons
"#;
    let cst = parse_or_panic(src);
    let zo = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(zo.clone());

    insta::assert_display_snapshot!(PrettyPrint(&zo));
}

#[test]
fn rev() {
    let src = r#"
ind(T: Set0) List
    case nil
    case cons(_: T, _: List)
    return Set0

fun rev(T: Set0, xs: List(T)): List(T)
    fun helper(-xs: List(T), acc: List(T)): List(T)
        match xs
        case nil:
            acc
        case cons(x, xs):
            helper(xs, cons(T)(x, acc))
        return1 List(T)
    helper(xs, nil(T))

ind Abc
    case a
    case b
    case c
    return Set0

let acons = cons(Abc)
let anil = nil(Abc)

let a_b_c = acons(a, acons(b, acons(c, anil)))

rev(Abc, a_b_c)
"#;
    let cst = parse_or_panic(src);
    let zo = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(zo.clone());

    insta::assert_display_snapshot!(PrettyPrint(&zo));
}

#[test]
fn eq() {
    let src = r#"
ind(T: Set0, left: T) Eq[_: T]
    case refl: [left]
    return Prop0
Eq
"#;
    let cst = parse_or_panic(src);
    let zo = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(zo.clone());

    insta::assert_display_snapshot!(PrettyPrint(&zo));
}

#[test]
fn eq_refl() {
    let src = r#"
ind(T: Set0, left: T) Eq[_: T]
    case refl: [left]
    return Prop0
refl
"#;
    let cst = parse_or_panic(src);
    let zo = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(zo.clone());

    insta::assert_display_snapshot!(PrettyPrint(&zo));
}

#[test]
fn eq_commutative() {
    let src = r#"
ind(T: Set0, left: T) Eq[_: T]
    case refl: [left]
    return Prop0

afun(T: Set0, a: T, b: T, eq: Eq(T, a)(b)): Eq(T, b)(a)
    match eq
    case refl:
        refl(T, a)
    use [c] return Eq(T, c)(a)
"#;
    let cst = parse_or_panic(src);
    let zo = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(zo.clone());

    insta::assert_display_snapshot!(PrettyPrint(&zo));
}

#[test]
fn eq_transitive() {
    let src = r#"
ind(T: Set0, left: T) Eq[_: T]
    case refl: [left]
    return Prop0

afun(
    T: Set0,
    a: T,
    b: T,
    c: T,
    ab: Eq(T, a)(b),
    bc: Eq(T, b)(c) // TODO: Support trailing commas.
): Eq(T, a)(c)
    let f =
        match ab
        case refl:
            afun(ac: Eq(T, a)(c)): Eq(T, a)(c)
                ac
        use [ina_outb]
        return For(_: Eq(T, ina_outb)(c)) -> Eq(T, a)(c)
    f(bc)
"#;
    let cst = parse_or_panic(src);
    let zo = may_to_zo(&cst).unwrap();

    assert_expr_is_well_typed_under_empty_tcon(zo.clone());

    insta::assert_display_snapshot!(PrettyPrint(&zo));
}
