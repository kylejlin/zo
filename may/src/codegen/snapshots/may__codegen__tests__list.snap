---
source: may/src/codegen/tests.rs
expression: PrettyPrint(&zo)
---
(
    fun
    nonrec
    (
        Type0
    )
    Type0
    (
        ind
        Type0
        "List"
        ()
        (
            (
                (
                    2
                    1
                )
                ()
            )
            (
                ()
                ()
            )
        )
    )
)
